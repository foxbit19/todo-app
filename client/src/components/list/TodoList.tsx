import React from 'react'
import { List } from '@mui/material'
import Item from '../../models/item'
import { DragDropContext, Droppable } from 'react-beautiful-dnd'
import TodoItem from '../item/TodoItem';

interface Props {
    items: Item[]
    onItemClick?: (item: Item) => void
    onComplete?: (item: Item) => void
}

const TodoList = (props: Props) => {
    return (
        <DragDropContext onDragEnd={() => { }}>
            <Droppable droppableId='1'>
                {(provided) => <List data-testid="list" ref={provided.innerRef} {...provided.droppableProps}>
                    {props.items.map((item: Item, index: number) => (
                        <TodoItem index={index} item={item}
                            onClick={() => props.onItemClick && props.onItemClick(item)}
                            onComplete={() => props.onComplete && props.onComplete(item)} />
                    ))}
                    {provided.placeholder}
                </List >}
            </Droppable>
            {props.items.length === 0 ? <div>There are no todos to show</div> : ''}
        </DragDropContext>
    )
}

export default TodoList
