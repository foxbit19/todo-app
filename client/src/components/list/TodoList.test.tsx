import React from 'react';
import { render, screen } from '@testing-library/react';
import Item from '../../models/item';
import sinon from 'sinon';
import TodoList from './TodoList';

describe('todo list', () => {
    const getItemElement = (itemId: number): HTMLInputElement => {
        return screen.getByTestId(`todo_${itemId}`)
    }

    test('it matches snapshot', () => {
        const view = render(<TodoList items={[]} />);
        expect(view).toMatchSnapshot();
    });

    test('it shows a message when the provided list of elements is empty', () => {
        render(<TodoList items={[]} />)

        expect(screen.getByText(/There are no todos to show/i)).toBeInTheDocument()
    })

    test('it shows the provided list of elements using description field', () => {
        const items: Item[] = [
            new Item(1, 'this is my first todo', 1),
            new Item(2, 'this is my second todo', 2),
            new Item(3, 'this is my third todo', 3)
        ]

        render(<TodoList items={items} />)

        for (const item of items) {
            expect(screen.getByText(item.description)).toBeInTheDocument()
        }
    })

    test('it truncates an item description if this text is longer than 50 characters', () => {
        const items: Item[] = [
            new Item(1, 'this is my first todo', 1),
            new Item(2, 'this is one of the longest todo that I\'ve wrote since then. This todo description is sooo long', 2),
        ]

        render(<TodoList items={items} />)

        for (const item of items) {
            const truncatedDescription = item.description.length > 50 ? `${item.description.substring(0, 50)}...` : item.description
            expect(screen.getByText(truncatedDescription)).toBeInTheDocument()
        }
    })

    test('it launch an event when an item is clicked', () => {
        const items: Item[] = [
            new Item(1, 'this is my first todo', 1),
            new Item(2, 'this is my second todo', 2),
        ]

        const handleItemClick = sinon.spy()

        render(<TodoList items={items} onItemClick={handleItemClick} />)
        screen.getByText(items[0].description).click()
        expect(handleItemClick.called).toBeTruthy()
    })

    test('it pass the item clicked on Item click', () => {
        const items: Item[] = [
            new Item(1, 'this is my first todo', 1),
            new Item(2, 'this is my second todo', 2),
        ]

        const handleItemClick = sinon.spy()

        render(<TodoList items={items} onItemClick={handleItemClick} />)
        screen.getByText(items[0].description).click()
        expect(handleItemClick.calledWithExactly(items[0])).toBeTruthy()
    })
})