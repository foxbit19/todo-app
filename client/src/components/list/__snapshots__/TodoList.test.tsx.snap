// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`todo list it matches snapshot 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        style="margin-top: 2em;"
      >
        <div
          class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-91853l-MuiGrid-root"
        >
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
              data-testid="InfoIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
              />
            </svg>
          </div>
          <div
            class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
          >
            <h6
              class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-alignCenter css-avn1fp-MuiTypography-root"
            >
              You can add, edit, delete and mark as complete an item. Drag an item over the list to change its priority.
            </h6>
          </div>
        </div>
        <h4
          class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter css-1nxg358-MuiTypography-root"
        >
          There are no todos to show
        </h4>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      style="margin-top: 2em;"
    >
      <div
        class="MuiGrid-root MuiGrid-container MuiGrid-spacing-xs-2 css-91853l-MuiGrid-root"
      >
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
            data-testid="InfoIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2zm1 15h-2v-6h2v6zm0-8h-2V7h2v2z"
            />
          </svg>
        </div>
        <div
          class="MuiGrid-root MuiGrid-item css-13i4rnv-MuiGrid-root"
        >
          <h6
            class="MuiTypography-root MuiTypography-subtitle2 MuiTypography-alignCenter css-avn1fp-MuiTypography-root"
          >
            You can add, edit, delete and mark as complete an item. Drag an item over the list to change its priority.
          </h6>
        </div>
      </div>
      <h4
        class="MuiTypography-root MuiTypography-h4 MuiTypography-alignCenter css-1nxg358-MuiTypography-root"
      >
        There are no todos to show
      </h4>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
